File: src/app/models/scooter.ts
/**
 * Enumeration for the scooter status
 */
// enum ScooterStatus {
//   IDLE = "Idle",
//   IN_USE = "In use",
//   MAINTENANCE = "Maintenance",
// }

export enum ScooterStatus {
  IDLE,
  IN_USE,
  MAINTENANCE
}

/**
 * Scooter class
 * @author Lawrance Bahem
 */
export class Scooter {
  private static _startingCode = 30001;
  private static _distanceLimit =  5000; // in meters
  private static _AmsterdamLocation = { latitude: 52.377956, longitude: 4.897070 };

  private _id: number;
  private _tag: String;
  private _status: ScooterStatus;
  private _gpsLocation: string;
  private _batteryCharge: number;
  private _setMileage: number;

  constructor(_id: number, _tag: String, _status: ScooterStatus, _gpsLocation: string, _mileage: number, _batteryCharge: number) {
    this._setMileage = _mileage;
    this._batteryCharge = _batteryCharge;
    this._gpsLocation = _gpsLocation;
    this._status = _status;
    this._tag = _tag;
    this._id = _id;
    this._tag = _tag;
    this._status = _status;
  }

  /**
   * Create a random scooter with random information
   */
  public static createRandomScooter(){
    const randomString = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
    const randomId = this._startingCode++;
    const randomTag = this.generateRandomTag(8, randomString);
    const randomStatus = this.getRandomStatus(ScooterStatus);
    const randomGPS = this.randomGeo(this._AmsterdamLocation.latitude,	this._AmsterdamLocation.longitude);
    const randomLocation = randomGPS.latitude + ',' + randomGPS.longitude;
    const randomMileage = randomGPS.distance;
    const batteryCharge = this.generateRandomPercentage();

    return new Scooter(randomId,randomTag, randomStatus, randomLocation, parseInt(randomMileage),batteryCharge);
  }

  /**
   * To generate random tag that consists of a certain length
   * @param lengthOfCode
   * @param possible
   * @private
   */
  private static generateRandomTag(lengthOfCode: number, possible: string) {
    let text = '';
    for (let i = 0; i < lengthOfCode; i++) {
      text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
  }

  /**
   * Generic method that returns a typesafe random enum.
   * @param enumOb
   * @private
   */
  private static getRandomStatus<T>(enumOb: T): T[keyof T] {
    const enumValues = Object.keys(enumOb)
      .map(n => Number.parseInt(n))
      .filter(n => !Number.isNaN(n)) as unknown as T[keyof T][];
    const randomIndex = Math.floor(Math.random() * enumValues.length);
    return enumValues[randomIndex];
  }

  /**
   * Generate random latitude, longitude. Based on a latitude and longitude
   * @param latitude
   * @param longitude
   * @private
   */
  private static randomGeo(latitude, longitude){
    const y0 = latitude;
    const x0 = longitude;
    const rd = this._distanceLimit / 111300; // about 111300 meters in one degree

    const u = Math.random();
    const v = Math.random();

    const w = rd * Math.sqrt(u);
    const t = 2 * Math.PI * v;
    const x = w * Math.cos(t);
    const y = w * Math.sin(t);

    const newLatitude = y + y0;
    const newLongitude = x + x0;
    const newDistance = this.distance(latitude, longitude, newLatitude, newLongitude).toFixed(2);
    return  {latitude : newLatitude, longitude: newLongitude, distance:newDistance};
  }

  /**
   * To calculate the distance between to coordinates
   * @param lat1 current
   * @param lon1 current
   * @param lat2 nearby
   * @param lon2 nearby
   * @private
   */
  private static distance(lat1, lon1, lat2, lon2):number{
    const R = 6371000;
    // tslint:disable-next-line:max-line-length
    const a = 0.5 - Math.cos((lat2 - lat1) * Math.PI / 180) / 2 + Math.cos(lat1 * Math.PI / 180)
      * Math.cos(lat2 * Math.PI / 180) * (1 - Math.cos((lon2 - lon1) * Math.PI / 180)) / 2;
    return Math.round(R * 2 * Math.asin(Math.sqrt(a)));
  }

  /**
   * To generate random number between 5-100
   * @private
   */
  private static generateRandomPercentage(): number{
    return Math.floor(5 + Math.random()*(100 + 1 - 5))
  }


  static get startingCode(): number {
    return this._startingCode;
  }


  static get distanceLimit(): number {
    return this._distanceLimit;
  }

  static set distanceLimit(value: number) {
    this._distanceLimit = value;
  }

  static get AmsterdamLocation(): { latitude: number; longitude: number } {
    return this._AmsterdamLocation;
  }

  static set AmsterdamLocation(value: { latitude: number; longitude: number }) {
    this._AmsterdamLocation = value;
  }

  get id(): number {
    return this._id;
  }

  set id(value: number) {
    this._id = value;
  }

  get tag(): String {
    return this._tag;
  }

  set tag(value: String) {
    this._tag = value;
  }

  get status(): ScooterStatus {
    return this._status;
  }

  set status(value: ScooterStatus) {
    this._status = value;
  }

  get gpsLocation(): string {
    return this._gpsLocation;
  }

  set gpsLocation(value: string) {
    this._gpsLocation = value;
  }

  get batteryCharge(): number {
    return this._batteryCharge;
  }

  set batteryCharge(value: number) {
    this._batteryCharge = value;
  }

  get setMileage(): number {
    return this._setMileage;
  }

  set setMileage(value: number) {
    this._setMileage = value;
  }

  getStatusToString(){
    return ScooterStatus[this.status];
  }

  getScooterStatusEnum(){
    return ScooterStatus;
  }

}

File: esserver/src/main/java/app/models/Scooter.java
package app.models;

import app.repositories.Identifiable;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonView;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;


@Entity
public class Scooter implements Identifiable {
  @Transient
  private final String randomString = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
  //  @JsonView({ShowScooterSummary.class, ShowId.class, ShowTag.class, ShowCharge.class, ShowStatus.class})
  @JsonView(ShowScooterSummary.class)
  @Id
  @GeneratedValue
  private int id;
  @JsonView(ShowScooterSummary.class)
  private String tag;
  @JsonView(ShowScooterSummary.class)
  @Enumerated(EnumType.STRING)
  private StatusScooter status;
  @JsonView(ShowScooterSummary.class)
  private int chargeBattery;
  @JsonIgnore
  public static int uniqueId = 30000;

  private String gpsLocation;
  private double mileage;

  @OneToMany(mappedBy = "scooter")
  private List<Trip> trips;
  @Transient
  Random random = new Random();
  public Scooter() {
    int randomIndex = random.nextInt(3);
    GeoLocation geoLocation = randomGeo(52.377956, 4.897070);
//      this.id = uniqueId++;
    this.status = StatusScooter.values()[randomIndex];
    this.tag = generateRandomTag(8, randomString);
    this.gpsLocation = geoLocation.lat + ", " + geoLocation.lon;
    this.chargeBattery = 0;
    this.mileage = geoLocation.distance;
    this.chargeBattery = generateRandomCharge();
    this.trips= new ArrayList<>();
  }

  public Scooter(String tag) {
    this();
    this.tag = tag;
  }

  public static Scooter creatRandomScooter() {
    return new Scooter();
  }

  public int generateRandomCharge() {
    return (int) Math.floor(5 + Math.random() * (100 + 1 - 5));
  }

  /**
   * Generate random GeoLocation
   *
   * @param latitude
   * @param longitude
   * @return
   */
  private GeoLocation randomGeo(double latitude, double longitude) {

    double y0 = latitude;
    double x0 = longitude;
    long rd = 5000 / 111300; // about 111300 meters in one degree

    double u = Math.random();
    double v = Math.random();

    double w = rd * Math.sqrt(u);
    double t = 2 * Math.PI * v;
    double x = w * Math.cos(t);
    double y = w * Math.sin(t);

    double newLatitude = y + y0+u;
    double newLongitude = x + x0+v;
    double newDistance = distance(latitude, newLatitude, longitude, newLongitude, "K");
    return new GeoLocation(newLatitude, newLongitude, newDistance);
  }


  /**
   * Calculate the distance between two distances
   *
   * @param lat1
   * @param lon1
   * @param lat2
   * @param lon2
   * @return
   */
  private double distance(double lat1, double lon1, double lat2, double lon2, String unit) {
    if ((lat1 == lat2) && (lon1 == lon2)) {
      return 0;
    } else {
      double theta = lon1 - lon2;
      double dist = Math.sin(Math.toRadians(lat1)) * Math.sin(Math.toRadians(lat2)) +
        Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) * Math.cos(Math.toRadians(theta));
      dist = Math.acos(dist);
      dist = Math.toDegrees(dist);
      dist = dist * 60 * 1.1515;
      if (unit.equals("K")) {
        dist = dist * 1.609344;
      } else if (unit.equals("N")) {
        dist = dist * 0.8684;
      }
      int random = (int) (Math.random() * 1000);
      return (double) (Math.round(dist * 100) / 100) + random;
    }
  }

  /**
   * To generate random tag
   *
   * @param lengthOfCode
   * @param possible
   * @return
   */
  private String generateRandomTag(int lengthOfCode, String possible) {
    StringBuilder text = new StringBuilder();
    for (int i = 0; i < lengthOfCode; i++) {
      text.append(possible.charAt((int) Math.floor(Math.random() * possible.length())));
    }
    return text.toString();
  }

  public Trip startNewTrip(LocalDateTime startDateTime){
    Trip trip = new Trip();
    if (this.status.equals(StatusScooter.IDLE)){
      trip.setStart(startDateTime);
      trip.setStartPosition(this.gpsLocation);
      String[] gpsLocation = this.gpsLocation.split(",");
      GeoLocation randomGeo = randomGeo(Double.parseDouble(gpsLocation[0]),Double.parseDouble(gpsLocation[1]));
      trip.setEndPosition(randomGeo.lat + ", " +randomGeo.lon);
      this.status = StatusScooter.IN_USE;

      return trip;
    }
    return null;
  }

  public void addTrip(Trip trip){
    this.trips.add(trip);
  }



  public String getTag() {
    return tag;
  }

  public String getGpsLocation() {
    return gpsLocation;
  }

  public int getChargeBattery() {
    return chargeBattery;
  }

  public double getMileage() {
    return mileage;
  }

  public StatusScooter getStatus() {
    return status;
  }


  public void setTag(String tag) {
    this.tag = tag;
  }

  @Override
  public int getId() {
    return this.id;
  }

  @Override
  public void setId(int id) {
    this.id = id;
  }

  public void setStatus(StatusScooter status) {
    this.status = status;
  }

  public void setChargeBattery(int chargeBattery) {
    this.chargeBattery = chargeBattery;
  }

  public static void setUniqueId(int uniqueId) {
    Scooter.uniqueId = uniqueId;
  }

  public void setGpsLocation(String gpsLocation) {
    this.gpsLocation = gpsLocation;
  }

  public void setMileage(double mileage) {
    this.mileage = mileage;
  }

  public void setTrips(List<Trip> trips) {
    this.trips = trips;
  }

  public void setRandom(Random random) {
    this.random = random;
  }

  public List<Trip> getTrips() {
    return trips;
  }




  /**
   * GeoLocation class to hold the random latitude and the longitude with the calculated distance
   * Between these coordinates and Amsterdam location
   */
  static class GeoLocation {
    private double lat;
    private double lon;
    private double distance;

    public GeoLocation(double lat, double lon, double distance) {
      this.lat = lat;
      this.lon = lon;
      this.distance = distance;
    }

    public double getLat() {
      return lat;
    }

    public double getLon() {
      return lon;
    }

    public double getDistance() {
      return distance;
    }

  }




   public enum StatusScooter {
    IDLE,
    IN_USE,
    MAINTENANCE
  }

  public class ShowScooterSummary {
  }

}
